### PR TITLE
CSE UI/UX: Animations & UI improvements

### PR DESCRIPTION
**Pull request by CZX123**
## Description
The PR introduces many new animations to the CSE machine and some UI improvements.

### UI Improvements
- Arrows from stash items are now always drawn as a straight line, and should appear less buggy now
- Colors in the CSE machine have changed slightly, especially for the print layout
  - New utility color functions created to represent different states or elements based on whether the CSE machine is in a print layout or not
  - Colors in the CSE machine for Java are also updated to use these new utility functions
- Array indexes now only show on hover or print layout, to reduce clutter
### Animation Changes
- Refactored animation components into two types:
  - one with the `animate` function, used for high-level animations with complex choreography
  - the other with the `animateTo` function where you can specify the target props, used in animations for simple nodes
- Reworked the base animation logic to use `Konva.Animation` instead of a `Konva.Tween` for much better flexibility
  - Added new `lerp` and `lerpColor` functions to interpolate between numbers and hex color strings for the animation logic to work
  - Multiple different animations can now run on the same animation component at the same time with different timings, by calling `animate` or `animateTo` in parallel
- Improved all of the existing animations, changing some timings and animating stroke colors as well
### New Animations
- Variable assignment
- Variable name lookup
- Environment instruction
- Branch instruction
- Frame creation (from block statements or function applications)
- Instruction application (more specifically, array literal instructions or built-in function call)
- Function application
- Array access instruction
- Array assignment

### CODE REVIEW

File: src/features/cseMachine/animationComponents/base/AnimationUtils.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+/** Omits the index signature `[key: string]: any;` from type `T` */
+export type RemoveIndex<T> = {
+  [K in keyof T as string extends K
+    ? never
+    : number extends K
+    ? never
+    : symbol extends K
+    ? never
+    : K]: T[K];
+};
```
Comment (RichDom2185) [inline src/features/cseMachine/animationComponents/base/AnimationUtils.ts:15]: Should we move this to `src/commons/utils/TypeHelpers.ts`?

File: src/features/cseMachine/animationComponents/FrameCreationAnimation.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+import { CseAnimation } from '../CseMachineAnimation';
+import { Config } from '../CseMachineConfig';
+import { ControlStashConfig } from '../CseMachineControlStashConfig';
+import { defaultActiveColor, defaultStrokeColor, isEnvEqual } from '../CseMachineUtils';
+import { Animatable, AnimationConfig } from './base/Animatable';
+import { AnimatedGenericArrow } from './base/AnimatedGenericArrow';
+import { AnimatedRectComponent, AnimatedTextComponent } from './base/AnimationComponents';
+import { getNodeDimensions, getNodeLocation, getNodePosition } from './base/AnimationUtils';
+
+export class FrameCreationAnimation extends Animatable {
```
Comment (RichDom2185) [inline src/features/cseMachine/animationComponents/FrameCreationAnimation.tsx:21]: Can we add a `/** ... */` comment here to help explain the animation for consistency with the other animation classes, but more importantly, for clarity?

File: src/features/cseMachine/animationComponents/FunctionApplicationAnimation.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  defaultStrokeColor,
+  getTextWidth
+} from '../CseMachineUtils';
+import { Animatable } from './base/Animatable';
+import { AnimatedGenericArrow } from './base/AnimatedGenericArrow';
+import { AnimatedTextbox } from './base/AnimatedTextbox';
+import { getNodeLocation, getNodePosition } from './base/AnimationUtils';
+import { FrameCreationAnimation } from './FrameCreationAnimation';
+
+export class FunctionApplicationAnimation extends Animatable {
```
Comment (RichDom2185) [inline src/features/cseMachine/animationComponents/FunctionApplicationAnimation.tsx:21]: Ditto on docs.

**Review by RichDom2185 — COMMENTED** [review]
Wow, I love this OOP design, seems very extensible. Not to mention the documentation comments making it clear, thanks for working on this!

I did a quick run through and here are some comments:

File: src/features/cseMachine/CseMachineAnimation.tsx [hunk 4]
```diff
@@ -25,99 +48,258 @@ export class CseAnimation {
...
   static updateAnimation() {
-    CseAnimation.animationComponents.forEach(a => a.destroy());
+    CseAnimation.animations.forEach(a => a.destroy());
     CseAnimation.clearAnimationComponents();
 
     if (!Layout.previousControlComponent) return;
     const lastControlItem = Layout.previousControlComponent.control.peek();
     const lastControlComponent = Layout.previousControlComponent.stackItemComponents.at(-1);
+    const currStashComponent = Layout.stashComponent.stashItemComponents.at(-1);
+    // const currControlComponent = Layout.controlComponent.stackItemComponents.at(-1);
```
Comment (RichDom2185) [inline src/features/cseMachine/CseMachineAnimation.tsx:145]: Is this used?

File: src/features/cseMachine/animationComponents/FrameCreationAnimation.tsx (hunk 2 above)
Comment (notnotmax) [reply src/features/cseMachine/animationComponents/FrameCreationAnimation.tsx:29]: Added a short description here, and same for the function application animation.

File: src/features/cseMachine/animationComponents/FunctionApplicationAnimation.tsx (hunk 3 above)
Comment (CZX123) [reply src/features/cseMachine/animationComponents/FunctionApplicationAnimation.tsx:27]: Updated

File: src/features/cseMachine/animationComponents/base/AnimationUtils.ts (hunk 1 above)
Comment (CZX123) [reply src/features/cseMachine/animationComponents/base/AnimationUtils.ts:15]: Didn't know about this file, updated in the latest commit

**Pull request merged**